### ISSUE TITLE
Different calls to dataSource function when scrolling or tapping

### ISSUE THREAD

**Issue by gmogames** [rechsteiner/Parchment#159]
Hi There,

Thank you so much for the awesome Lib.
I have a question related using it as a DataSource and when tapping to select an option or when scrolling between controllers.

This is my datasource implementation:
```
func pagingViewController<T>(_ pagingViewController: PagingViewController<T>, pagingItemForIndex index: Int) -> T {
        guard let pagingIndex = PagingIndexItem(index: index, title: tabs[index].name()) as? T else {
            return PagingIndexItem(index: index, title: "") as! T // force unwrap needed for generic return
        }
        return pagingIndex
    }
    
    func pagingViewController<T>(_ pagingViewController: PagingViewController<T>, viewControllerForIndex index: Int) -> UIViewController {
        let tab = tabs[index]
        return tab.mappedViewController()
    }
    
    func numberOfViewControllers<T>(in: PagingViewController<T>) -> Int {
        return tabs.count
    }
```

I was testing a weird issue happening, on the viewcontroller I have a collectionview with a custom header and in that header I have a filter menu. I noticed that when I chose one of the menu options and only swipped between controllers, the menu and data never changed, but if I tapped on the tab option, the menu and data would refresh as a fresh new load of the view.

I then put a breakpoint in pagingViewController:viewControllerForIndex and noticed that when I swiped from one controller to the other, that function got called once, but when I tapped on a tab, it gets called 3 times and the index changes the 3 times.

I'm not sure I'm doing this correctly. I'm trying to make sure the state of the VC remain the same independently of being scrolled into or tapped, but I cant seem to figure out why the behavious is different.

Video:
![out](https://user-images.githubusercontent.com/1765046/38006629-bb6745b8-31fa-11e8-816d-7db9f7308553.gif)

Any input would be highly appreciated.

Thank you so much

**Comment by rechsteiner** [comment]
Thanks for the report! Added a fix for this in #160 ✌️ You have to target the `master` branch for now, but I'll release a new version after easter. Let me know you're still seeing any issues.

**Comment by gmogames** [comment]
Perfect. Thank you very much!

**Comment by gmogames** [comment]
@rechsteiner I also noted that when I scroll from one tab to the other, the willScrollToItem delegate gets called right, but when I tap on an item, this delegate does not get called. But the didScrollToItem delegate gets called on a tap and on a scroll.

I'm not sure this is the intended behavior, but it seems that all delegates should be called on tap or scroll otherwise if you have logic that depend on it, you will be in trouble (as I am right now 😆 )

**Comment by rechsteiner** [comment]
Yeah, that definitely sounds like a bug. Will look into fixing that 👍

**Comment by gmogames** [comment]
@rechsteiner sorry, another one for you.
I noticed that if I start scrolling to the next/previous tab but release the scroll, going back to where I was, the didScrollToItem gets called with the destinationViewController of the tab I was scrolling to but never did.

So if i gently start swiping to the next tab and release, that gets called.

In my case, I have a search delegate that needs to be set to the viewController that was scrolled to, so my delegate was changing to the destinationViewController but the user was still on the same tab, and now without the delegate, with search not working.

I'm assuming the didScrollToItem should only be called if the scroll finished and the current selected tab is different then the one the user was.

Thanks!

**Comment by rechsteiner** [comment]
Fixed the issue with the willScrollToItem delegate now: #173. Thanks for catching that 👌 Will try to get a new version out today!

As for the other issue, I agree that it's a bit weird that didScrollToItem is called even if you cancel a transition. It's based on the delegates that I get from `EMPageViewController`, but I will look into changing this in the future. This will be a breaking API change, so it's not something I will change before releasing version 2.0.

**Comment by rechsteiner** [comment]
This was released in version 1.2.0. Let me know if you're seeing any other issues!
